src/digest.rs

```diff
@@ -448,8 +447,66 @@ pub extern fn SHA512_4(out: *mut u8, out_len: c::size_t,
     polyfill::slice::fill_from_slice(out, digest);
 }
 
+unsafe extern fn sha1_block_data_order(state: *mut u64, data: *const u8, num: c::size_t) {
+    sha1_process_blocks(
+        &mut *(state as *mut _),
+        core::slice::from_raw_parts(data as *const _, num)
+    )
+}
+
+type Sha1State = [u32; 5];
+type Sha1Block = [u32; 16];
+
+fn sha1_process_blocks(state: &mut Sha1State, blocks: &[Sha1Block]) {
+    #[inline] fn ch(x: u32, y: u32, z: u32) -> u32 { (x & y) | (!x & z) }
+    #[inline] fn parity(x: u32, y: u32, z: u32) -> u32 { x ^ y ^ z }
+    #[inline] fn maj(x: u32, y: u32, z: u32) -> u32 { (x & y) | (x & z) | (y & z) }
+
+    let mut w: [u32; 80] = [0; 80];
+    for block in blocks {
+        for t in 0..16 {
+            w[t] = u32::from_be(block[t])
+        }
+        for t in 16..80 {
+            w[t] = (w[t - 3] ^ w[t - 8] ^ w[t - 14] ^ w[t - 16]).rotate_left(1)
+        }
+
+        let mut a = state[0];
+        let mut b = state[1];
+        let mut c = state[2];
+        let mut d = state[3];
+        let mut e = state[4];
+
+
+        for t in 0..80 {
+            let (k, f) = match t {
+                0...19 => (0x5a827999, ch(b, c, d)),
+                20...39 => (0x6ed9eba1, parity(b, c, d)),
+                40...59 => (0x8f1bbcdc, maj(b, c, d)),
+                60...79 => (0xca62c1d6, parity(b, c, d)),
+                _ => unreachable!()
+            };
+            let tt = a.rotate_left(5)
+                .wrapping_add(f)
```

Comment:
Nit: Is this wrapped correctly? It looks pretty horrific to me. It seems like it would be better to align the dots.

Also, wouldn't it be better to use `Wrapping<u32>` instead, since all the math here is wrapping? It seems like it would be less error -prone to use `Wrapping<u32>` and also easier to read.